### ISSUE TITLE
Serious performance issue repeatedly adding single features

### ISSUE THREAD

**Issue by westnordost** [tangrams/tangram-es#1135]
I have a serious performance issue in my [StreetComplete app](https://github.com/westnordost/StreetComplete), using tangram 0.4.5.

**Short explanation of what I do**
I have a lot of map pins saved in a local database. I pull these selectively from DB, depending on in which tile the user is currently viewing the map at, [using my own slippy map math](https://github.com/westnordost/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/tangram/QuestsMapFragment.java#L170).
The pins found in the displayed area are then added to the MapView once via `layer.addPoint(pos, props)`. The class memorizes for which tiles it already pulled the pins from DB so if the user scrolls by that area later, the pins are not re-retrived from DB.

**The issue**
Analyzing the performance problem, I drilled down to that above 90% of the whole time taken for the process is spent in `MapData::addPoint(pos, props)`. I cannot measure deeper than that without adding time measurements in native code.
To give you an idea of dimension, I am trying to display around 300-500 pins on the map, which takes about 15-40 seconds(!). Everything else (determine area to pull, getting pins from db, creating a string map for tags, calling `applySceneUpdates` happens in below 1s).

Also, this could be a valuable observation:
For the first few dozens of points, calling `addPoint` finishes very fast (0-5ms), but at a certain point, it jumps up and takes 10-40ms  per call. If `addPoint `is called even more often, at some point it takes ~200ms per call. I guess the performance ever so worsens the more often I call the method (repeatedly).

**Reproduction**
So in short, to reproduce, add time measurements to the `addPoint` (or in general, add map feature) method and call this method ~1000 times. Observe how the performance drops dramatically.

**Comment by lunarflint** [comment]
After some testing it is one of these two lines (or both) slow down the things.
```
    std::lock_guard<std::mutex> lock(m_mutexStore);
    m_store = std::make_unique<GeoJSONVT>(m_features, m_maxZoom, m_maxZoom, indexMaxPoints, tolerance);
```
You can take a look the following methods at clientGeoJsonSource.cpp  (git checkout tags/v0.4.5-android)
```
void ClientGeoJsonSource::addData(const std::string& _data)
void ClientGeoJsonSource::addPoint(const Properties& _tags, LngLat _point)
void ClientGeoJsonSource::addLine(const Properties& _tags, const Coordinates& _line)
void ClientGeoJsonSource::addPoly(const Properties& _tags, const std::vector<Coordinates>& _poly)
```

**I suggest two ways to solve this:**
- move that 3 lines (along with the line `m_generation++;`) out and form a `update()` method, expose it to the java wrapper. And we have to call that method every time we add something, for instance, after a for-loop of mapData.addPoint(). This way requires the least efforts.
- create methods like addPoints(), addPolygons(), etc.

And for now if you just need a quick fix, you can just convert all the points to a GeoJson and call mapData.addGeoJson(str);

**Little bit offtopic stuff:**
I am using tangram-es for similar data illustration purpose and I need to load and hot-swap data real quick. Seems that there are many useful api on the C++ lib that are not exposed to Java. For example, I took a look at TileManager and found this:
```
    /* Returns the set of currently visible tiles */
    const auto& getVisibleTiles() { return m_tiles; }
```
I have not tried this out yet. But you may take a look at it, so you don't have to calculate it by yourself again.

**Comment by westnordost** [comment]
If I understood the documentation for `std::make_unique` correctly, it constructs a new `GeoJSONVT` and thus, more importantly, copies the whole list of `m_features` since it not a reference or pointer. So for every feature added, the whole list of features is copied. I guess the immediate problem here is that `GeoJSONVT` as the underlying data structure is not designed to be mutable. I imagine this would be quite an effort to change.

Taking a few steps back from the immediate problem, the use case of @lunarflint and me should be a quite common one: To have a data source where the map pulls the data from by itself instead of pushing the data onto the map "by hand" (calling `addPoint` etc).
I think a good final solution, at least from a Java standpoint, would be to have an interface similar to this
```
interface MapDataSource {
  Collection<MapFeature> get(BoundingBox bbox);
}
``` 
and add a method like `MapController::addDataSource(String name, MapDataSource dataSource)` to the MapController. Tangram-ES then manages pulling the data from the specified data source by visible tiles itself and the user of the library just has to implement that interface - behind the interface, there could be a database, a static list of features, a http client talking to a REST API, ... all the same.

To support mutable data sources (like I use/need them), there'd need to be an interface like
```
interface MutableMapDataSource extends MapDataSource {
  void addMapFeature(MapFeature featuer);
  // and probably:
  void removeMapFeature(MapFeature feature);
  // and perhaps addPoint, addPolyline etc.
}
``` 
With the logic to talk to and handle these interfaces in place, it is a very small step to efficiently support adding single features to a standard `MapData` layer. The MapData layer would just be a simple data structure that implements MutableMapDataSource.

**Comment by lunarflint** [comment]
hmm, I tried out the `update()` thing myself. It takes less than 0.5sec to add 1000 polygons.

**Comment by blair1618** [comment]
Hello, thanks for the report and investigation! This problem has been noted before (https://github.com/tangrams/tangram-es/pull/847), and it sounds like a fix is needed pretty badly now. 

The crux of the problem, as you've figured out, is that the interface for `GeoJSONVT` only allows us to submit a single collection of features and receive the tiled results - we can't incrementally update the set of features. This means that when you add a feature, we have to re-generate the collection of all the features in that layer. Naturally, this gets slower when the number of features in the layer becomes large. 

One way to improve this is to allow `MapData` to accept bulk collections of features, rather than forcing you to add one feature at a time. 

Another thing that would help is to bucket your features into more layers (i.e. more `MapData` objects). This would make each single update depend on fewer features. 

To move forward with revising the `MapData` interface to allow bulk feature submission, what methods or behaviors would you find most helpful?

**Comment by westnordost** [comment]
@lunarflint suggested to _"create methods like `addPoints()`, `addPolygons()`, etc."_, I find this the most sensible immediate feature because it wouldn't be a breaking change, unlike his other suggestion.

Now the question is how "broken" `addPoint`, `addPolygon` etc. you consider in their current implementation state. If you want to force users to rethink their usage of these methods, implementing his first suggestion (`update()` method) will break their code, making them acquainted with the issue.

Both our use cases suggest that as soon as a user would want to add geometry in bulk, the thing he actually wants to achieve is to have a custom data source, ideally somehow managed by tangram-es. That was my suggestion.
Realizing that this will probably much more work to do and needs to be properly planned, IMO methods like `addPoints()`, `addPolygons()` etc will do a good job of relieving this issue now.

**Comment by lunarflint** [comment]
@blair1618:
Talking about buckets, the problem is I don't really know how many buckets I should have. For my use case, it would be extremely beneficial if I can add/remove/modify an arbitrary number of sub layers on the fly. The top layer contains the style but no data, and the children link to a MapData class. I know it is not possible now and I am happy with hard-coding buckets.

Btw, I really appreciate the good work of everyone on this project.

@westnordost 
To minimize the amount of calls to GeoJSONVT, maybe this is even better (just like the SharedPreferences class of Android SDK):
```
MapData.edit().addPolygon(...).addPoint(...).addPoint(...).commit();
```

**Comment by westnordost** [comment]
Since you just released the marker API in 0.4.6, perhaps here the question: Are the markers intended to be added in huge amounts? I.e. is the data structure in which they are kept eligible for that?

**Comment by matteblair** [comment]
Resolved by #2043
